### PR TITLE
[WIP] Change manually created queries using Knex.js API

### PR DESCRIPTION
**Pull request by maturanomx**
My PR is a: 💅 Enhancement
Main update on the: Plugin

**NOTE**: This is a **WORK IN PROGRESS**. I created this PR for visibility of the work that I have plan to do in next days for possible early comments and the opportunity of keep track of the evolution. Not working full time on this, so could be take some time :upside_down_face: 

---

## Description
As mentioned in #1462, I'm starting to work on adding support for MSSQL database (a personal requirement). Pursuing the best approach and good code quality, I'll create different PRs with gradual changes seeking a easy code review.

**The scope of this PR is just refactor the actual implementation** on `strapi-hook-bookshelf` without changing his behavior. The objective here is avoid the queries manually created and increase the use of Knex.js

## Pros
With using Knex.js --**a dependency that Strapi already have**-- it's expected obtain the benefits of a **dedicated query builder**: the work of more number and experienced eyes.

Avoiding reinventing the wheel, we could reduce the lines of code<sup id="rf1">[[1]](#fn1)</sup>, with that reducing the maintenance costs in this project and the keep the space for bugs at the minimum.

Also, this will open the opportunity to create better schemas for relational databases, like add real constrains and real foreign keys<sup id="rf2">[[2]](#fn2)</sup>, increasing their performance and data integrity. 

## Cons
_TBD_

## What is tested
- [ ] Creation of the initial schema whit the creation of a new project
- [ ] Case of use when using UUID as primary key
- [ ] Creating a Content-Type
- [ ] All possible data-types for a Content Type
- [ ] Adding fields to an already create

## Results
_TBD_

---
###### Footnotes
[1](#rf1). <span id="fn1"></span> To be verified. I have not done the work, is just a possibility that I picture in my mind.
[2](#rf2). <span id="fn2"></span> Out of the scope of this PR. It just to mention the opportunity of improvement.

### CODE REVIEW

**Comment by derrickmehaffy** [comment]
@maturanomx 
```
on update current_timestamp trigger removed from updated_at
```
Very important on MySQL as this basically tells the server to update only that field and not created_at when new information is "PUT"

**Comment by derrickmehaffy** [comment]
@maturanomx 

```
Changes longtext fields types by text type
```

I made that change because @soupette is storing some **extremely** large entires in core_store

https://github.com/strapi/strapi/pull/1643

**Comment by derrickmehaffy** [comment]
Something to be looked in with more detail is backend support for defining custom "accuracy" one some fields like float, double, decimal, ect

`All possible data-types for a Content Type`

**Comment by derrickmehaffy** [comment]
@maturanomx Is this something that can easily be included as a patch for boolean values?

https://github.com/strapi/strapi/issues/1925

**Comment by derrickmehaffy** [comment]
Also per previous question wanted to see if this could also be included: https://github.com/strapi/strapi/issues/1330

**Comment by derrickmehaffy** [comment]
Related to Keys: https://github.com/strapi/strapi/issues/1189

**Comment by maturanomx** [comment]
First at all, thanks @derrickmehaffy for keeping and eye on this 😃, your feedback is really important.

> `on update current_timestamp trigger removed from updated_at`
>
> Very important on MySQL as this basically tells the server to update only that field and not created_at when new information is "PUT"

My question is, why is this trigger [excluded for PostgreSQL](https://github.com/strapi/strapi/blob/6a24021a2a88830ecbeb42bf73ebffda5864d353/packages/strapi-hook-bookshelf/lib/index.js#L432). One of ~~two~~ three options:
1. Is not necessary. The update is done by code somewhere 🤔 ...
2. We have a bug in PostgreSQL. The column `updated_at` is not been updated on modifications 😱 
3. Somewhere in code, the update is done just for PostgreSQL; doing the opposite of what is done here 🤦‍♂️ 

... I know you are an experiment MySQL user, if you said if required now I will make the changes to get the original behavior. I will take note for the PostgreSQL case.

**Comment by derrickmehaffy** [comment]
Yeah not sure on postgres, it should update when PUT requests are done, I know on MySQL that is being handled correctly, using that means that Strapi doesn't need to send any formatted date codes because the DB server handles it on its own.

Based on a bit of googling around: https://x-team.com/blog/automatic-timestamps-with-postgresql/

**Comment by maturanomx** [comment]
> > Changes longtext fields types by text type
>
> I made that change because @soupette is storing some extremely large entires in core_store
>
> #1643 

Question: What are the implications of manage just `longtext` columns?. Could not have quicker searches in small/adequate column sizes? Is not the size of the database bigger or consumes more memory or something like that??? (honest question 😄)

I'm thinking in another approach: make the distinction in the model definition.
```
// core_store model definition file
"attributes": {
  "some_column": "longtext", // Or "mediumtext" "text"
}
```

MySQL could take that value, other databases just create "text" as now. Another benefit of this approach if the possibility of create an UI according the text-size...

What do you thing?

File: packages/strapi-hook-bookshelf/lib/index.js [hunk 1]
```diff
@@ -528,14 +528,75 @@ module.exports = function(strapi) {
...
+                          // Common case, numeric incrementable key
+                          dbTable.increments();
+
+                        } else {
+                          // Build primary key with a peculiar field type
+                          const pk = dbTable[definition.primaryKeyType]('id')
+                            .primary()
+                            .notNullable();
+
+                          if (definition.client === 'pg' && definition.primaryKeyType === 'uuid') {
```
[the remark this answers is not in the document]
Comment (maturanomx) [reply packages/strapi-hook-bookshelf/lib/index.js]: My idea is first replace the current behavior. I have read other "possible improvements" that you mention (please, continue) and I think we could add it --like this, is possible--, but I'm thinking in let those for other PR for a better review and inclusion.

I'm letting commends in code in the meantime 😄

File: packages/strapi-hook-bookshelf/lib/index.js [hunk 2]
```diff
@@ -533,14 +533,84 @@ module.exports = function(strapi) {
...
+                                  // ROOM FOR IMPROVEMENT -- Remove distinction; procur same behavior on all engines
+                                  definition.client === 'mysql' && field.type === 'timestampUpdate'
+                                    ? ORM.knex.raw('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP')
+                                    : ORM.knex.fn.now()
+                                );
+                                break;
+
+                              case 'enumeration':
+                                // ROOM FOR IMPROVEMENT -- Fix from model definition
+                                dbTable.enu(fieldName, field.enum);
```
Comment (derrickmehaffy) [inline packages/strapi-hook-bookshelf/lib/index.js]: Should use varchar, enum data type on MySQL is terrible

File: packages/strapi-hook-bookshelf/lib/index.js [hunk 3]
```diff
@@ -533,14 +533,84 @@ module.exports = function(strapi) {
...
+                              // Creating integer column for foreign key as the original implementation
+                              // ROOM FOR IMPROVEMENT -- Use actual foreign keys
+                              field.model && dbTable.integer(field.model);
+                              return;
+                            }
+
+                            switch(field.type) {
+                              // ROOM FOR IMPROVEMENT -- We could add REAL defined constrains to database
+                              case 'email':
+                              case 'password':
```
Comment (derrickmehaffy) [inline packages/strapi-hook-bookshelf/lib/index.js]: password isn't be hidden by default

**Comment by derrickmehaffy** [comment]
To break down our slack convo:

```
create table `test1s` (`id` int unsigned not null auto_increment primary key, `teststring` varchar(255), `testint` int, `testfloat` float(8, 2), `testdec` decimal(8, 2), `testdate` timestamp default current_timestamp, `testemail` varchar(255), `testenum` enum('test1', 'test2', 'test3'), `testtext` text, `testboolean` boolean, `testjson` json, `testpassword` varchar(255), `created_at` timestamp default current_timestamp, `updated_at` timestamp default current_timestamp on update current_timestamp) default character set utf8;
```

Float/Decimal should have user defined accuracy (above the scope of this PR I think, but needs to be done before merging)

The date field (manual user field) shouldn't be current_timestamp

enum shouldn't use the MySQL data type, use varchar as enum on MySQL is a huge PITA

Text fields should use longtext by default (same with json, since in order to use json data type you need at least mysql v8 or mariadb v10.2 and ubuntu 18.04 ships with mariadb v10.1 by default)

created_at should use current_timestamp (which it does)
updated_at should use on update current_timestamp

**Comment by maturanomx** [comment]
## Diff from latest version 
_Spaces added, when possible, just for better visualization_

### MySQL:
```diff
- create table `testings` ( id  int          auto_increment not null primary key, `field_string` varchar(255), `field_integer` int, `field_float` double     , `field_decimal` decimal(10, 2), `field_date` timestamp default current_timestamp, `field_date_withdefault` timestamp default current_timestamp, `field_email` varchar(255), `field_enumeration` varchar(255), `field_text` longtext, `field_boolean` boolean, `field_json` longtext, `field_password` varchar(255), `created_at` timestamp default current_timestamp, `updated_at` timestamp default current_timestamp on update current_timestamp);
+ create table `testings` (`id` int unsigned not null auto_increment primary key, `field_string` varchar(255), `field_integer` int, `field_float` float(8, 2), `field_decimal` decimal(10, 2), `field_date` timestamp default current_timestamp, `field_date_withdefault` timestamp default current_timestamp, `field_email` varchar(255), `field_enumeration` varchar(255), `field_text` longtext, `field_boolean` boolean, `field_json` longtext, `field_password` varchar(255), `created_at` timestamp default current_timestamp, `updated_at` timestamp default current_timestamp on update current_timestamp) default character set utf8;

- create table `core_store` ( id  int          auto_increment not null primary key, `key` varchar(255), `value` longtext, `type` varchar(255), `environment` varchar(255), `tag` varchar(255));
+ create table `core_store` (`id` int unsigned not null auto_increment primary key, `key` varchar(255), `value` longtext, `type` varchar(255), `environment` varchar(255), `tag` varchar(255)) default character set utf8;

- create table `upload_file` ( id  int          auto_increment not null primary key, `name` varchar(255), `hash` varchar(255), `sha256` varchar(255), `ext` varchar(255), `mime` varchar(255), `size` varchar(255), `url` varchar(255), `provider` varchar(255), `created_at` timestamp default current_timestamp, `updated_at` timestamp default current_timestamp on update current_timestamp);
+ create table `upload_file` (`id` int unsigned not null auto_increment primary key, `name` varchar(255), `hash` varchar(255), `sha256` varchar(255), `ext` varchar(255), `mime` varchar(255), `size` varchar(255), `url` varchar(255), `provider` varchar(255), `created_at` timestamp default current_timestamp, `updated_at` timestamp default current_timestamp on update current_timestamp) default character set utf8;

- create table `users-permissions_permission` ( id  int          auto_increment not null primary key, `type` varchar(255), `controller` varchar(255), `action` varchar(255), `enabled` boolean, `policy` varchar(255), `role` integer);
+ create table `users-permissions_permission` (`id` int unsigned not null auto_increment primary key, `type` varchar(255), `controller` varchar(255), `action` varchar(255), `enabled` boolean, `policy` varchar(255), `role` int) default character set utf8;

- create table `users-permissions_role` ( id  int          auto_increment not null primary key, `name` varchar(255), `description` varchar(255), `type` varchar(255));
+ create table `users-permissions_role` (`id` int unsigned not null auto_increment primary key, `name` varchar(255), `description` varchar(255), `type` varchar(255)) default character set utf8;

- create table `users-permissions_user` ( id  int          auto_increment not null primary key, `username` varchar(255), `email` varchar(255), `password` varchar(255), `resetpasswordtoken` varchar(255), `confirmed` boolean, `blocked` boolean, `role` integer);
+ create table `users-permissions_user` (`id` int unsigned not null auto_increment primary key, `username` varchar(255), `email` varchar(255), `password` varchar(255), `resetpasswordtoken` varchar(255), `confirmed` boolean, `blocked` boolean, `role` int) default character set utf8;

- create table `upload_file_morph` ( id  int          auto_increment not null primary key, `upload_file_id` int, `related_id` int, `related_type` longtext, `field` longtext);
+ create table `upload_file_morph` (`id` int unsigned not null auto_increment primary key, `upload_file_id` int, `related_id` int, `related_type` longtext, `field` longtext) default character set utf8;
```

### PostgreSQL:
```diff
- create table "upload_file" ( id  serial primary key, "name" varchar(255), "hash" varchar(255), "sha256" varchar(255), "ext" varchar(255), "mime" varchar(255), "size" varchar(255), "url" varchar(255), "provider" varchar(255), "created_at" timestamp with time zone, "updated_at" timestamp with time zone);
+ create table "upload_file" ("id" serial primary key, "name" varchar(255), "hash" varchar(255), "sha256" varchar(255), "ext" varchar(255), "mime" varchar(255), "size" varchar(255), "url" varchar(255), "provider" varchar(255), "created_at" timestamptz default current_timestamp, "updated_at" timestamptz default current_timestamp);

- create table "users-permissions_permission" ( id  serial primary key, "type" varchar(255), "controller" varchar(255), "action" varchar(255), "enabled" boolean, "policy" varchar(255), "role" integer);
+ create table "users-permissions_permission" ("id" serial primary key, "type" varchar(255), "controller" varchar(255), "action" varchar(255), "enabled" boolean, "policy" varchar(255), "role" integer);

- create table "users-permissions_user" ( id  serial primary key, "username" varchar(255), "email" varchar(255), "password" varchar(255), "resetpasswordtoken" varchar(255), "confirmed" boolean, "blocked" boolean, "role" integer);
+ create table "users-permissions_user" ("id" serial primary key, "username" varchar(255), "email" varchar(255), "password" varchar(255), "resetpasswordtoken" varchar(255), "confirmed" boolean, "blocked" boolean, "role" integer);

- create table "users-permissions_role" ( id  serial primary key, "name" varchar(255), "description" varchar(255), "type" varchar(255));
+ create table "users-permissions_role" ("id" serial primary key, "name" varchar(255), "description" varchar(255), "type" varchar(255));

- create table "testings" ( id  serial primary key, "field_string" varchar(255), "field_integer" integer, "field_float" double precision, "field_decimal" decimal(10, 2), "field_date" timestamp with time zone, "field_date_withdefault" timestamp with time zone, "field_email" varchar(255), "field_enumeration" varchar(255), "field_text" text, "field_boolean" boolean, "field_json" jsonb, "field_password" varchar(255), "created_at" timestamp with time zone, "updated_at" timestamp with time zone);
+ create table "testings" ("id" serial primary key, "field_string" varchar(255), "field_integer" integer, "field_float" real, "field_decimal" decimal(10, 2), "field_date" timestamptz default current_timestamp, "field_date_withdefault" timestamptz default current_timestamp, "field_email" varchar(255), "field_enumeration" varchar(255), "field_text" text, "field_boolean" boolean, "field_json" jsonb, "field_password" varchar(255), "created_at" timestamptz default current_timestamp, "updated_at" timestamptz default current_timestamp);

- create table "core_store" ( id  serial primary key, "key" varchar(255), "value" text, "type" varchar(255), "environment" varchar(255), "tag" varchar(255));
+ create table "core_store" ("id" serial primary key, "key" varchar(255), "value" text, "type" varchar(255), "environment" varchar(255), "tag" varchar(255));

- create table "upload_file_morph" ( id  serial primary key, "upload_file_id" integer, "related_id" integer, "related_type" text, "field" text);
+ create table "upload_file_morph" ("id" serial primary key, "upload_file_id" integer, "related_id" integer, "related_type" text, "field" text);
```

**Comment by maturanomx** [comment]
@derrickmehaffy 

> Float/Decimal should have user defined accuracy (above the scope of this PR I think, but needs to be done before merging)

Why? Originally is not user defined, right? I'm seeking now having the same behavior  for have the safest merge possible. Right now I just have a difference for the `float` file type. Can you help me to investigate/test how important is this?

> enum shouldn't use the MySQL data type, use varchar as enum on MySQL is a huge PITA
>
> Text fields should use longtext by default (same with json, since in  order to use json data type you need at least mysql v8 or mariadb v10.2  and ubuntu 18.04 ships with mariadb v10.1 by default)
>
> created_at should use current_timestamp (which it does)
> updated_at should use on update current_timestamp

done, done, done & done :+1: 

---
> The date field (manual user field) shouldn't be current_timestamp

This...

I would think this is how it should be, and I tried to remove this (you can see, **currently have this default value**) but I got an strange error:
screenshot 2018-09-19 23 23 22

Honestly, I don't know the reason of this error :man_shrugging:. Why I can create the first one without a default value but need one for the second? :thinking:. The query for your test:
```sql
CREATE TABLE `test` (`field_date` timestamp, `field_date_two` timestamp);
``` 
 
Besides of knowing the problem, change the current behavior is out of scope of this PR. When is merged, adding a default user-selected value, instead of the current one could be done. But, that's work for future...

File: packages/strapi-hook-bookshelf/lib/index.js (hunk 3 above)
Comment (maturanomx) [reply packages/strapi-hook-bookshelf/lib/index.js]: Do you mean in the GUI? ... or not been hashed in the DB?
If is something of that, should not be caused by this changes. The same behavior should be present in `master`. If not, can you expand what are you meaning please?

**Comment by TokenChingy** [comment]
@derrickmehaffy 
> Float/Decimal should have user defined accuracy (above the scope of this PR I think, but needs to be done before merging)

Firstly, user defined floats and decimals is a bad idea. It's better to standardise, we should agree the context of decimals and floats.

Decimals in Strapi should be defined as the conventional monetary convention of a float that is accurate to up to 2 decimal places.

Floats in Strapi should be as precise as possible (on the Mongoose ORM are defined as a Decimal128 to up to 20 decimal positions — when the fix is merged in 3.0.0-alpha14.1.2), natively JS handles floats as with a 64-bit precision.

**Comment by derrickmehaffy** [comment]
@maturanomx @TokenChingy I disagree that user defined values for float is a bad idea. Decimal makes sense. I spoke at length with @lauriejim on this issue in the past. There are cases where a user will need a certain number of decimal points. Realistically if you need extremely high accuracy on mysql you should be using the `double` data type (which it currently does). Mongo AFAIK doesn't support the typical schema for defining high accuracy floating point types.

Since Strapi is a framework, allowing user choice is important. I have stated this to @Aurelsicoko as well that yes Strapi should be simple, but you shouldn't lock out your advanced users for the sake of simplicity. Everyone will have their own use case for this framework and we should appeal to both novice and advanced users.

As for actually defining it, it is honestly quite simple instead of just just using `float` you use `float(23,20)`

To be completely honest I personally think these types are something that should be defined based on the database.

IF Mongo enabled then use these data types (x,y,z)
IF MySQL/PG enabled then use these data types (a,b,c,d,e,f,g)

For instance with the INT in mysql allows for the following range: `-2147483648 to 2147483647` however if you define it as unsigned then you get the following range: `0 to 4294967295`. Depending on a users usecase they should have the option to choose. In this case lets say they have a serial number they need to store, that serial number will never be a negative number. Meaning without the choice their serials cannot exceed 2,147,483,647. However unlikely it will be, they should still have that choice and same applies to decimal, float, and double.

There are many cases where Strapi doesn't use both Mongo and MySQL to it's full potential. But these additions to data types should be fairly easy, it really only requires a few more options on the advanced tab while creating the value and reading those variables.

**Comment by derrickmehaffy** [comment]
That is until the user can define a "custom" data type which has been requested in the past.

**Comment by derrickmehaffy** [comment]
Here is an easy list of MySQL data types:
https://www.techonthenet.com/mysql/datatypes.php

Postgres: https://www.techonthenet.com/postgresql/datatypes.php

**Comment by derrickmehaffy** [comment]
Here is an example @maturanomx for datetime

```
ALTER TABLE `cmdrs`
	ADD COLUMN `test` DATETIME NULL DEFAULT NULL AFTER `updated_at`;
```

The problem is that strapi sends it in some odd format with the timezone MySQL is extremely strict that the format must be `YYYY-MM-DD HH:MM:SS`

**Comment by Aurelsicoko** [comment]
I'm blown away by the work done here... Thank you so much @maturanomx, @derrickmehaffy and @TokenChingy for your contribution to the project ❤️

**Comment by maturanomx** [comment]
Adding new fields to an already create model, done. Originally was executing a query by new field; this implementation is doing all in just one query.

```diff
- ALTER TABLE "foos" ADD "field_integer" integer;
- ALTER TABLE "foos" ADD "field_float" double precision;
- ALTER TABLE "foos" ADD "field_date" timestamp with time zone;
- ALTER TABLE "foos" ADD "field_email" varchar(255);
- ALTER TABLE "foos" ADD "field_enumeration" varchar(255);
- ALTER TABLE "foos" ADD "field_boolean" boolean;
- ALTER TABLE "foos" ADD "field_password" varchar(255);
+ alter table "foos" add column "field_integer" integer, add column "field_float" real, add column "field_date" timestamptz default CURRENT_TIMESTAMP, add column "field_email" varchar(255), add column "field_enumeration" varchar(255), add column "field_boolean" boolean, add column "field_password" varchar(255)
```

**Comment by maturanomx** [comment]
@derrickmehaffy @Aurelsicoko @lauriejim

This PR has been ready to be merge since a long time ago. I'm going to resume work for support MS-SQL, if you can help me reviewing and testing this changes in order to can continue with that...

Thanks.

**Comment by derrickmehaffy** [comment]
@maturanomx I will do some very detailed testing to make sure everything performs as it should and let you know.

File: packages/strapi-hook-bookshelf/lib/index.js [hunk 4]
```diff
@@ -523,99 +432,126 @@ module.exports = function(strapi) {
...
+                              : dbTable.text(fieldName, 'longtext');
+                            break;
+
+                          case 'email':
+                          case 'password':
+                          case 'enumeration': // Manage native type is too complicated
+                            column = dbTable.string(fieldName);
+                            break;
+
+                          case 'date':
```
Comment (derrickmehaffy) [inline packages/strapi-hook-bookshelf/lib/index.js]: Missing datetime:

`case 'datetime':`

**Comment by derrickmehaffy** [comment]
@maturanomx see the linked PR about date, I think some modifications should be made to include those new dates at least for MySQL.

File: packages/strapi-hook-bookshelf/lib/index.js [hunk 5]
```diff
@@ -417,10 +416,10 @@ module.exports = function(strapi) {
                             type = definition.client === 'pg' ? 'integer' : 'int';
                             break;
                           case 'float':
-                            type = definition.client === 'pg' ? 'double precision' : 'double';
+                            type = definition.client === 'pg' ? 'double precision' : 'double     ';
```
Comment (derrickmehaffy) [inline packages/strapi-hook-bookshelf/lib/index.js]: Not sure why there is spaces after double?

**Comment by maturanomx** [comment]
@derrickmehaffy @lauriejim ... There was a problem because [I'm mirroring this repository on Gitlab](https://gitlab.com/maturanomx/strapi) and the branch for this PR wasn't updated there... I fixed that.

sorry for the absence, I was assigned as team leader of an urgent project and is tooking all my time. Hope back soon...

**Comment by derrickmehaffy** [comment]
Just wanted to reiterate this @maturanomx @lauriejim 

At least with MySQL:
`datetime` should be used when defined by the user in a model
`timestamp` should be used with `created_at` (auto generated by strapi)
`timestampUpdate` should be used with `updated_at` (auto generated by strapi)

See the following explanation here:
```
The DATETIME type is used when you need values that contain both date and time information. MySQL retrieves and displays DATETIME values in 'YYYY-MM-DD HH:MM:SS' format. The supported range is '1000-01-01 00:00:00' to '9999-12-31 23:59:59'.

...

The TIMESTAMP data type has a range of '1970-01-01 00:00:01' UTC to '2038-01-09 03:14:07' UTC. It has varying properties, depending on the MySQL version and the SQL mode the server is running in.
```
https://dev.mysql.com/doc/refman/8.0/en/datetime.html

**Comment by derrickmehaffy** [comment]
^^^ That is until the changes that @johannpinson suggested on adding more date/time/datetime fields are added.

**Comment by derrickmehaffy** [comment]
@maturanomx If you don't have time to test this completely, and get ready to merge it I can do some testing and repush the PR?

**Comment by derrickmehaffy** [comment]
I'm currently testing the following changes with the date/time/datetime parts, right now the `date` option can't be used in the adminUI but it would be just the date without any time data.

```javascript
                          case 'date':
                            column = dbTable.date(fieldName);
                            break;

                          case 'datetime':
                            column = dbTable.datetime(fieldName);
                            break;

                          case 'timestamp':
                            column = dbTable.timestamp(fieldName).defaultTo(ORM.knex.fn.now());
                            break;

                          case 'timestampUpdate':
                            column = dbTable.timestamp(fieldName).defaultTo(
                              // ROOM FOR IMPROVEMENT -- Remove distinction; procur same behavior on all engines
                              definition.client === 'mysql' && field.type === 'timestampUpdate'
                                ? ORM.knex.raw('CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP')
                                : ORM.knex.fn.now()
                            );
                            break;
```
